include/can/core/dispatch.hpp

```diff
@@ -0,0 +1,131 @@
+        // Sort for binary search
+        std::sort(arr.begin(), arr.end());
+    }
+
+    /**
+     * Look up message id in collection
+     * @param id message id
+     * @return True if present
+     */
+    [[nodiscard]] auto in(can_ids::MessageId id) const -> bool {
+        return std::binary_search(arr.cbegin(), arr.cend(), id);
+    }
+
+  private:
+    std::array<can_ids::MessageId, sizeof...(T)> arr{};
+};
+
+/**
+ * Concept describing a message handling type.
+ * @tparam T The type of the class
+ * @tparam Message The message types it handles
+ */
+template <typename T, typename... Message>
+concept HandlesMessages =
+    requires(T t, std::variant<std::monostate, Message...>& message) {
+    // Has a handle method that accepts a variant.
+    {t.handle(message)};
+};
+
+/**
+ * A CanMessageBufferListener that parses messages and notifies a Handler
+ * @tparam HandlerType A HandlesMessages type
+ * @tparam MessageTypes The Message types this handles
+ */
+template <typename HandlerType, CanMessage... MessageTypes>
+requires HandlesMessages<HandlerType, MessageTypes...>
+class DispatchParseTarget {
+  public:
+    DispatchParseTarget(HandlerType& handler) : handler{handler}, parser{} {}
+
+    template <bit_utils::ByteIterator Input, typename Limit>
+    requires std::sentinel_for<Limit, Input>
+    void handle(uint32_t arbitration_id, Input input, Limit limit) {
+        auto arb = ArbitrationId{.id = arbitration_id};
+        auto result =
+            parser.parse(MessageId{arb.parts.message_id}, input, limit);
+        handler.handle(result);
+    }
+
+  private:
+    HandlerType& handler;
+    Parser<MessageTypes...> parser;
+};
+
+/**
+ * A CanMessageBufferListener that will write messages to a MessageBuffer only
+ * if arbitration id matches message ids in MessageTypes
+ *
+ * @tparam BufferType a MessageBuffer
+ * @tparam MessageTypes The message types
+ */
+template <message_buffer::MessageBuffer BufferType,
+          HasMessageID... MessageTypes>
+class DispatchBufferTarget {
+  public:
+    explicit DispatchBufferTarget(BufferType& buffer)
+        : writer{buffer}, coll{} {}
+
+    template <bit_utils::ByteIterator Input, typename Limit>
+    requires std::sentinel_for<Limit, Input>
+    void handle(uint32_t arbitration_id, Input input, Limit limit) {
+        auto arb = ArbitrationId{.id = arbitration_id};
+        if (coll.in(can_ids::MessageId{arb.parts.message_id})) {
+            writer.send(arbitration_id, input, limit, 100);
+        }
+    }
+
+  private:
+    CanMessageBufferWriter<BufferType> writer;
+    MessageIdCollection<MessageTypes...> coll;
+};
+
+/**
+ * A CanMessageBufferListener that will dispatch messages to other
+ * CanMessageBufferListeners
+ * @tparam Listener CanMessageBufferListener objects
+ */
+template <CanMessageBufferListener... Listener>
+class Dispatcher {
+  public:
+    explicit Dispatcher(Listener*... listener) : registered{listener...} {}
```

Comment:
what's the expected lifetime of these listener pointers? should they be references instead?